xonsh/base_shell.py

```diff
@@ -148,7 +148,7 @@ def default(self, line):
         tee = Tee() if store_stdout else io.StringIO()
         try:
             ts0 = time.time()
-            self.execer.exec(code, mode='single', glbs=self.ctx)  # no locals
+            run_code_with_cache(src, self.execer, self.ctx, None, 'single')
```

Comment:
Does `self.push(line)` above not always compile the code? It looks to me that it will now always be compiled at least once. It kind of ruins the point of a cache. Sorry if there is something that I completely miss. I am not familiar with this part of the code..